### PR TITLE
[cmake] switch to CMake's native `check_{compiler,linker}_flag`

### PR DESCRIPTION
**Pull request by h-vetinari**
Broken out from #93429

Somewhat closing the loop opened by https://github.com/llvm/llvm-project/commit/7017e6c9cfd2de3122ce9528f338a97d61e96373

### CODE REVIEW

**Comment by jhuber6** [comment]
Here's a radical question, do we really want to use CMake's support for this? I remember a discussion recently about the increasingly large amount of time spent in the CMake configuration step, and most of that time is spent during these flag checks which pretty much all compile + link some file with no parallelism. I've also had issues working with these flags when trying to cross-compile things for the GPU, namely because the compilation flags insist on checking the linker so I need to do something like `set(CMAKE_REQUIRED_FLAGS "-c -flto")` to prevent it from invoking non-LLVM binaries for NVIDIA compilation.

File: cmake/Modules/LLVMCheckCompilerLinkerFlag.cmake [hunk 1]
```diff
@@ -1,35 +0,0 @@
...
-  # have succeeded if it was included on both commands.
-  #
-  # Therefore, try adding the flag to CMAKE_REQUIRED_FLAGS, which gets
-  # added to both compiling and linking commands in the tests.
-
-  cmake_push_check_state()
-  set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} ${flag}")
-  if(COMMAND check_compiler_flag)
-    check_compiler_flag("${lang}" "" ${out_var})
-  else()
```
Comment (h-vetinari) [inline cmake/Modules/LLVMCheckCompilerLinkerFlag.cmake:23]: see here...

File: llvm/cmake/modules/LLVMCheckLinkerFlag.cmake [hunk 2]
```diff
@@ -1,28 +0,0 @@
-include(CheckLinkerFlag OPTIONAL)
-
-if (COMMAND check_linker_flag)
-  macro(llvm_check_linker_flag)
-    check_linker_flag(${ARGN})
-  endmacro()
-else()
```
Comment (h-vetinari) [inline llvm/cmake/modules/LLVMCheckLinkerFlag.cmake:7]: ... and here

**Review by h-vetinari — COMMENTED** [review]
> Here's a radical question, do we really want to use CMake's support for this?

I think this is a fair question in the big picture, but the current state of the code is _already_ using CMake for that (see below) - I'm just removing the fallbacks for old CMake versions.

IMO this PR should stay a trivial clean-up, and any other flag-check-refactorings should come on top.

**Comment by ldionne** [comment]
The Bootstraping build seems to be failing with a real error: https://github.com/llvm/llvm-project/actions/runs/9596576098/job/26465655299?pr=96171

**Review by arichardson — COMMENTED** [review]
Since I believe the CMake minimum version supports this, I don't see why we shouldn't submit this cleanup. Whether to change the flag detection logic can be a separate discussion IMO.

**Comment by h-vetinari** [comment]
The failing tests here previously were clearly unrelated (missing python module `packaging` somewhere). The failure I see now is due to a timeout (also unrelated):
```
llvm-lit: /home/runner/_work/llvm-project/llvm-project/llvm/utils/lit/lit/main.py:72: note: The test suite configuration requested an individual test timeout of 0 seconds but a timeout of 1500 seconds was requested on the command line. Forcing timeout to be 1500 seconds.
-- Testing: 9661 tests, 30 workers --
  interrupted by user, skipping remaining tests
```

CI hasn't finished yet, so there may be others things hiding still, but I don't consider it realistic (to be caused by the changes here). The code before/after this PR should work exactly as before, we're just stripping the fallbacks for old CMake.

**Comment by h-vetinari** [comment]
Failing bootstrap build still has:
```
    from packaging import version
ModuleNotFoundError: No module named 'packaging'
```
which is clearly unrelated to this PR.

**Comment by ldionne** [comment]
@h-vetinari This one still looks like a real issue: https://buildkite.com/llvm-project/github-pull-requests/builds/74912#01903ca1-8089-4388-afdb-b65d400e6315

The rest looks good indeed.

**Comment by h-vetinari** [comment]
I think that was the last one. :)

**Comment by ldionne** [comment]
I'm sorry to be the bearer of bad news, but it looks like Android is failing with this patch: https://buildkite.com/llvm-project/libcxx-ci/builds/35981#01905269-56a0-4314-a350-0b99ece37e89

This is kinda weird, it fails due to the ABI list changing. I checked other recent executions on Android and it doesn't look spurious. @rprichard is responsible for the Android CI bot and may be able to help.

**Comment by rprichard** [comment]
Sorry, I didn't notice the ping until just now.

The CMake feature testing seems broken, on Android at least. Here's a representative error from `build/android-ndk-21-def-x86/CMakeFiles/CMakeError.log`:

```
Performing C++ SOURCE FILE Test CXX_SUPPORTS_NOSTDLIBXX_FLAG failed with the following output:
Change Dir: /llvm/build/android-ndk-21-def-x86/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/local/bin/ninja cmTC_80129 && [1/2] Building CXX object CMakeFiles/cmTC_80129.dir/src.cxx.o
[2/2] Linking CXX executable cmTC_80129
FAILED: cmTC_80129 
: && /opt/android/clang/clang-current/bin/clang++ --target=i686-linux-android21 --sysroot=/opt/android/ndk/sysroot --start-no-unused-arguments --unwindlib=none --end-no-unused-arguments  CMakeFiles/cmTC_80129.dir/src.cxx.o -o cmTC_80129   && :
ld.lld: error: unable to find library -lc++
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.


Source file was:
int main() { return 0; }
```

It's trying to see if `-nostdlib++` is supported, but I don't see that flag being passed in the `clang++` command line that CMake uses. In any case, CMake tries  to build a C++ program, which fails because libc++ doesn't exist yet.

It seems that the Android support in the Dockerfile is broken after a refactor last year. I have some changes that get it working again. BuildKite is still successfully running Android tests as of today, so I think it's using an old image.

**Comment by h-vetinari** [comment]
> It seems that the Android support in the Dockerfile is broken after a refactor last year. I have some changes that get it working again.

 I was waiting for your recent PRs to land, but the android failures seem to persist even after them.

**Comment by rprichard** [comment]
> I was waiting for your recent PRs to land, but the android failures seem to persist even after them.

Yeah that's also what I noticed. My PRs were aimed at getting the Android Dockerfile buildable again, but once it was buildable, the feature testing failed.

It _seemed_ that after this PR, the CMake feature test for a flag `-foo` wasn't even passing `-foo` to the Clang driver, but maybe I was reading too much into the CMake log output. It definitely wasn't handling the bootstrapping situation correctly -- it was trying to detect whether the driver supported `-nostdlib++` and concluded that it didn't, because it couldn't link an executable using `-lc++` (because libc++ hadn't been built yet).

My suspicion was that this PR broke CMake feature testing for any cross-compiled configurations.

**Comment by rprichard** [comment]
This check in libcxx/cmake/config-ix.cmake seems to break:
```
check_cxx_compiler_flag(-nostdlib++ CXX_SUPPORTS_NOSTDLIBXX_FLAG)
```

It appears that it isn't passing either `-nostdlib++` or `-lc++` to the clang linker command-line, but then `-lc++` _is_ passed to `ld.lld` because `-nostdlib++` is missing. The check fails because libc++ hasn't been built yet.

It seems that the older function had been dealing with this:

```
-function(llvm_check_compiler_linker_flag lang flag out_var)
-  # If testing a flag with check_c_compiler_flag, it gets added to the compile
-  # command only, but not to the linker command in that test. If the flag
-  # is vital for linking to succeed, the test would fail even if it would
-  # have succeeded if it was included on both commands.
-  #
-  # Therefore, try adding the flag to CMAKE_REQUIRED_FLAGS, which gets
-  # added to both compiling and linking commands in the tests.
...
```

It sounds like the right fix is to use `check_linker_flag` though? I can test that...

**Comment by rprichard** [comment]
> This check in libcxx/cmake/config-ix.cmake seems to break:
> 
> ```
> check_cxx_compiler_flag(-nostdlib++ CXX_SUPPORTS_NOSTDLIBXX_FLAG)
> ```

This particular check was already using `check_cxx_compiler_flag` before this PR, so it was broken(?), but it apparently doesn't matter because the real check is in runtimes/CMakeLists.txt instead, and that was previously using `llvm_check_compiler_linker_flag`.

Not sure how the runtimes build influences the libcxx/libcxxabi builds. It feels like `CXX_SUPPORTS_NOSTDLIBXX_FLAG` from the runtimes build is being inherited?

**Comment by rprichard** [comment]
The previous `llvm_check_compiler_linker_flag` passes the same flag to both the compile step and the linker step. Apparently, it's like a combined `check_compiler_flag` and `check_linker_flag`. CMake apparently doesn't implement this directly (https://stackoverflow.com/questions/78503180/cmake-how-to-check-for-a-compiler-and-linker-flag-at-the-same-time), but it's not too hard to implement it ourselves. The LLVM implementation is much better than the one proposed on StackOverflow.

If we just want to do a cleanup, then `llvm_check_compiler_linker_flag` can be simplified a lot:

```cmake
include(CMakePushCheckState)
include(CheckCompilerFlag)

function(llvm_check_compiler_linker_flag lang flag out_var)
  # If testing a flag with check_compiler_flag, it gets added to the compile
  # command only, but not to the linker command in that test. If the flag
  # is vital for linking to succeed, the test would fail even if it would
  # have succeeded if it was included on both commands.
  #
  # Therefore, try adding the flag to CMAKE_REQUIRED_FLAGS, which gets
  # added to both compiling and linking commands in the tests.

  cmake_push_check_state()
  set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} ${flag}")
  check_compiler_flag("${lang}" "" ${out_var})
  cmake_pop_check_state()
endfunction()
```

Almost all the uses of `llvm_check_compiler_linker_flag`  that https://github.com/llvm/llvm-project/pull/96171/commits/1515c2cd16911e1297c57fb69d4bb4921a40b8ab switches to `check_compiler_flag` are actually testing a linker flag, so they need to use `check_linker_flag` instead.

There were a few flags that I wasn't familiar with, so maybe they could be `check_compiler_flag`, or maybe we actually need to do both:

 * Darwin's `-fapplication-extension`
 * `/SAFESEH` to MSVC in openmp/runtime/cmake/config-ix.cmake
 * ICC's `-static-intel`
 * ICC's `-no-intel-extensions`

Ignoring those cases, the Android tests pass with this PR after switching `llvm_check_compiler_linker_flag` call sites to `check_linker_flag`.

**Comment by rprichard** [comment]
I rebased the changes so I could review them, https://github.com/llvm/llvm-project/commit/640d2e631a4bc11e3afa030e30d05288ee954cb5.

The compiler-rt/cmake/config-ix.cmake file is still removing `include(LLVMCheckCompilerLinkerFlag)`, but it needs to keep it now.

**Comment by h-vetinari** [comment]
> I rebased the changes so I could review them, https://github.com/llvm/llvm-project/commit/640d2e631a4bc11e3afa030e30d05288ee954cb5.

The github UI should show all changes to main? Not sure why a rebase would be necessary...?

> The compiler-rt/cmake/config-ix.cmake file is still removing `include(LLVMCheckCompilerLinkerFlag)`, but it needs to keep it now.

That's because it is included twice (see first & last line below):
https://github.com/llvm/llvm-project/blob/8dd574236ccaa0a183278396cfec3068b832651a/compiler-rt/cmake/config-ix.cmake#L3-L8

**Comment by h-vetinari** [comment]
If you're interested in everything but the first commit, the GH UI can do that too (there's a dropdown in the upper lefthand corner):
https://github.com/llvm/llvm-project/pull/96171/files/1df587efeb71fb1929667f008d7e9b251863d9d8..a47554828053c039916b005f15446df14f594e50

**Comment by rprichard** [comment]
> The github UI should show all changes to main? Not sure why a rebase would be necessary...?

Sorry, you're right. (Not sure why I didn't notice that before.)

The change looks good to me now.

**Review by ldionne — APPROVED** [review]

**Comment by rorth** [comment]
This patch broke the [Solaris/amd64](https://lab.llvm.org/staging/#/builders/120/builds/485) and [Solaris/sparcv9](https://lab.llvm.org/buildbot/#/builders/13/builds/1287) buildbots.

**Comment by ldionne** [comment]
@rorth Please check if this is still broken after https://github.com/llvm/llvm-project/commit/7ab643383fe49d5bb22e45610cb52476bf7a922e

**Comment by rorth** [comment]
Both bots are working again: thanks for the quick fix.

**Pull request merged**